Design of data preprocessing in AutoML.md

```diff
@@ -0,0 +1,93 @@
+
+### Approaches to consider
+(1) *PipelineModel approach*
+
+Similar to MojoPipelineBuilder we can have PipelineModelBuilder. It will have ability to add main model that is supposed to perform final scoring and also an ability to add preprocessing models ( keys to those trained models).
+
+We will also at some point need something like `MappingSpec` that we already have on mojo side, to be able to specify generated by preprocessors column names. But ideally we should have an ability to specify output columns for all our models in the first place, to avoid renaming in PipelineModel and to avoid dependencies on current naming conventions of existing preprocessors ( e.g. TE adds "_te" suffix for encoded columns)
+
+PipelineModel model could be seen as a proxy to a main scoring model. Even name on leaderboard should reflect this fact - keep original name of model (maybe just add suffix "_pipeline" that signifies that preprocessing was used).
+
+#### Pros:
+-  will contribute to a simmetry between native and mojo/production aspects of the codebase. In production we use MojoPipeline to fullfil corresponding task
+- will probably provide better isolation in terms of how many classes we will have to "touch" or affect
+- as pipeline is a separate concept it will be a place for doing pipeline related logic, e.g. how sequence of the preprocessing models is going to be applied, how data flows from preprocessor to preprocessor( input and outputs). I believe that this functionality is not Model specific and could/should be abstracted out.
+
+#### Cons:
+- wrapping original main model into PipelineModel will hide its functionality. It can be mitigated by introducing ProxyModel interface with `getMainModel` method that will provide main model to the user. Though, for the clients (theoretically) PipelineModel could be flattenned automatically. Complexity of doing this  depends on how flexible we are with serialisation at rest api layer.
+
+![PipelineModel approach](pipelinemodel.png)
+
+(2) *Integration preprocessing functionality into a Model class*
+
+We can probably have two sub-approaches here:
+- when we have PipelineModelBuilder and during training it collapses into main model without any wrapper around it
+- when we extend functionality of ModelBuilder by adding ability to have references to proprocessing models
+
+Not sure which approach is better but probably we can avoid introduction of the PipelineModelBuilder class. But it will increase number of responsibilities of the ModelBuilder class.
+
+#### Pros:
+- We do not hide any functionality behind PipelineModel wrapper
+- All models could be equally seen as pipelines ( which could also be a cons.)
+
+#### Cons:
+- conceptually abusing notion of the Model. Model ( what is also known as Transformer in other frameworks) is supposed to be only trained on data and after that be able to transform similar data. Nothing about data preprocessing is expected here. See Concerns section for more details.
```

Comment:
Only if you focus on generic preprocessing. If we stay in the context of Target Encoding, it fits the Model very well because that is what the Model does for all other types of encodings.

Also having it built-in has some inherent advantages. The encoding can be applied just to the columns the model actually decides to use.

Reply:
@michalkurka 
>because that is what the Model does for all other types of encodings.

this was explained further in Concerns section. I was talking here in the most general sense. Models are not supposed to do encoding or any preprocessing on their own. But in current state of h2o-3's design  Model class does these kind of things. In that sense, yes, hardcoding TE inside Model will fits well. In my opinion it is not an optimal long-term direction. Also embedding TE in a non generic way(like other existing encoders) will be a contribution to an already existing debt as hardcoded solutions are not scalable ones.

> Also having it built-in has some inherent advantages. The encoding can be applied just to the columns the model actually decides to use.

 If by "model decides" you mean which columns are going to be excluded due to "bad columns" and "ignore columns" mechanisms I would say that having them first encoded will prevent Model from "deciding" to ignore those columns. In short, model's decisions should come after data preprocessing steps.

Anyway I think that both approaches actually are in equal position in that concern.